lib/base/dependencygraph.hpp

```diff
@@ -21,12 +23,85 @@ class DependencyGraph
 	static void AddDependency(Object *parent, Object *child);
 	static void RemoveDependency(Object *parent, Object *child);
 	static std::vector<Object::Ptr> GetParents(const Object::Ptr& child);
+	static std::vector<Object::Ptr> GetChildren(const Object::Ptr& parent);
 
 private:
 	DependencyGraph();
 
+	/**
+	 * Represents an undirected dependency edge between two objects.
+	 *
+	 * It allows to traverse the graph in both directions, i.e. from parent to child and vice versa.
+	 */
+	struct Edge
+	{
+		// Note, due the existing confusing implementation of the DependencyGraph class, the terms parent and
+		// child are flipped here. Meaning, the parent is the dependent object and the child is the dependency.
+		// Example: A service object is dependent on a host object, thus the service object is the *parent* and
+		// the host object is the *child* one :).
+		Object* parent;
+		Object* child;
+		// Is used to track the number of dependent edges of the current one.
+		// A number <= 1 means, this isn't referenced by anyone and might soon be erased from the container.
+		// Otherwise, each remove operation will only decrement this by 1 till it reaches 1 and causes the
+		// edge to completely be erased from the container.
+		int count;
+
+		Edge(Object* parent, Object* child, int count = 1): parent(parent), child(child), count(count)
+		{
+		}
+
+		struct HashPredicate
+		{
+			/**
+			 * Generates a unique hash of the given Edge object.
+			 *
+			 * Note, the hash value is generated only by combining the hash values of the parent and child pointers.
+			 *
+			 * @param edge The Edge object to be hashed.
+			 *
+			 * @return size_t The resulting hash value of the given object.
+			 */
+			size_t operator()(const Edge& edge) const
+			{
+				size_t seed = 0;
+				boost::hash_combine(seed, edge.parent);
+				boost::hash_combine(seed, edge.child);
+
+				return seed;
+			}
+
+			/**
+			 * Compares whether the two Edge objects contain the same parent and child pointers.
+			 *
+			 * Note, the member property count is not taken into account for equality checks.
+			 *
+			 * @param a The first Edge object to compare.
+			 * @param b The second Edge object to compare.
+			 *
+			 * @return bool Returns true if the two objects are equal, false otherwise.
+			 */
+			bool operator()(const Edge& a, const Edge& b) const
+			{
+				return a.parent == b.parent && a.child == b.child;
+			}
+		};
+	};
+
+	using DependencyMap = boost::multi_index_container<
+		Edge, // The value type we want to sore in the container.
+		boost::multi_index::indexed_by<
+			// The first indexer is used for lookups by the Edge object itself, thus it
+			// needs its own hash function and comparison predicate.
+			boost::multi_index::hashed_non_unique<boost::multi_index::identity<Edge>, Edge::HashPredicate, Edge::HashPredicate>,
```

Comment:
"by the Edge object itself" is a bit vague as `Edge` includes `count` which isn't included in the lookup. I'd rather say  something like "the edge from child to parent".